### PR TITLE
[Relay] Remove overwriting of matmul shapes when they are static

### PR DESCRIPTION
**Pull request by jwfromm**
In the Relay Matmul shape relation, we are a little over enthusiastic about unifying dynamic shapes. If one of the shapes is static, it does not need to be unified. This change only rewrites dynamic shapes to required static constraints.

### CODE REVIEW

File: src/relay/op/nn/nn.h [hunk 1]
```diff
@@ -115,20 +115,36 @@ bool MatmulRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
         auto sb = B_shape.size();
         if (transpose_a && transpose_b) {
           auto tmp = A_shape[sa - 2];
-          A_shape[sa - 2] = B_shape[sb - 1];
-          B_shape[sb - 1] = tmp;
+          if (A_shape[sa - 2].as<tir::AnyNode>()) {
```
Comment (AndrewZhaoLuo) [inline src/relay/op/nn/nn.h:118]: There is now a ton of nesting. To make it readable I suggest a structure like 
```
A_shape_new = ...
B_shape_new = ...
index_swap_a, index_swap_b;

if (transpose_a && transpose_b) {
   // calculate index_swap_a and index_swap_b
} ... 

if (A_shape[index_swap_a].as<Any>...) {
    // do swap
}

if (B_shape[index_swap_b].as<Any>...
```

**Review by AndrewZhaoLuo — APPROVED** [review]
Thanks, one nit.

File: src/relay/op/nn/nn.h (hunk 1 above)
Comment (jwfromm) [reply src/relay/op/nn/nn.h]: Yeah good call, updated in latest push.

**Review by slyubomirsky — CHANGES_REQUESTED** [review]
The changes seem reasonable, but I think we should add at least some test cases that demonstrate the difference in logic to avoid possible regression.

**Comment by jwfromm** [comment]
I've added a check to `test_dense` to confirm that the input shapes aren't unnecessary made dynamic that fails without this change.

**Review by slyubomirsky — APPROVED** [review]
Thanks for adding in the assertion. I'm surprised this quirk was lurking in an existing test case.

**Review by shingjan — APPROVED** [review]
LGTM. Thanks for adding the test!

**Pull request merged**